File: Data Structures & Algorithms/matrixDFS/submission-0.cpp
class Solution {
public:
    int countPaths(vector<vector<int>>& grid) {
        unordered_set<int> visited;
        return backtrack(grid, visited, 0, 0);
    }

private:
    int backtrack(vector<vector<int>>& grid, unordered_set<int>& visited, int r, int c) {
        if (min(r, c) < 0 || r >= grid.size() || c >= grid[0].size()) {
            return 0;
        }
        if (grid[r][c] == 1) {
            return 0;
        }
        int position = getPosition(grid.size(), r, c);
        if (visited.find(position) != visited.end()) {
            return 0;
        }
        if (r == grid.size() - 1 && c == grid[0].size() - 1) {
            return 1;
        }
        int count = 0;
        visited.insert(position);
        count += backtrack(grid, visited, r + 1, c);
        count += backtrack(grid, visited, r - 1, c);
        count += backtrack(grid, visited, r, c + 1);
        count += backtrack(grid, visited, r, c - 1);
        visited.erase(position);
        return count;
    }

    int getPosition(int ROW, int r, int c) {
        return r * ROW + c;
    }
};

File: Data Structures & Algorithms/count-number-of-islands/submission-0.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid or not grid[0]:
            return 0
        res = 0
        visited = set() # (x, y)

        m, n = len(grid), len(grid[0])
        for r in range(m):
            for c in range(n):
                if grid[r][c] == '1' and (r, c) not in visited:
                    res += 1
                    self.traverse(grid, visited, r, c, m, n)
        
        return res
    
    def traverse(self, grid, visited, r, c, m, n):
        if r < 0 or r == m or c < 0 or c == n:
            return
        if grid[r][c] != '1':
            return
        if (r, c) in visited:
            return
        
        visited.add((r, c))
        self.traverse(grid, visited, r + 1, c, m, n)
        self.traverse(grid, visited, r - 1, c, m, n)
        self.traverse(grid, visited, r, c + 1, m, n)
        self.traverse(grid, visited, r, c - 1, m, n)
